publish to ESA as a independend processing step

nicHoch opened the pull request:
add a persistent file (sqlite) where we keep track of what has been delivered to ESA so fare.

with the storage in place we can check for certain conflicts and come up with rules how to resolve.

nicHoch committed:
first version publish step

nicHoch committed:
add version and product filter

nicHoch committed:
fix call error

versioning

move esa duplicates to folder

add scp method to publish step

add documentation and sub1/2 strategy

add doc

first version publish step

add version and product filter

fix call error

versioning

move esa duplicates to folder

add scp method to publish step

add documentation and sub1/2 strategy

add doc

add BSD request reason LUT handling
TODO: connect to API

connect to platform request reason LUT endpoint

samaloney commented on stixcore/processing/publish.py:
```diff
@@ -0,0 +1,300 @@
+                       f"D {len(files[PublishConflicts.SAME_ESA_NAME])} "
+                       f"A {len(files[PublishConflicts.ADDED])}")
+            error = "" if 'failed' not in files else "ERROR-"
+            message = f"""Subject: StixCore TMTC Publishing {error}Report {summary}
+
+ERRORS
+******
+{files['failed']}
+
+PUBLISHED DUPLICATES
+********************
+{pformat(files[PublishConflicts.SAME_EXISTS])}
+
+PUBLISHED ESA DUPLICATES
+************************
+{pformat(files[PublishConflicts.SAME_ESA_NAME])}
+
+PUBLISHED
+*********
+{pformat(files[PublishConflicts.ADDED])}
+
+
+StixCore
+
+==========================
+
+do not answer to this mail.
+"""
```
move to var defined at top of file may `PUBLISH_EMAIL_TEMPLATE = f""" ...."""`

nicHoch replied:
good point

samaloney commented on stixcore/processing/publish.py:
```diff
@@ -0,0 +1,300 @@
+
+
+if __name__ == '__main__':
+    publish_fits_to_esa(sys.argv[1:])
```
So would the plan to be to run this a cron job or similar?

nicHoch replied:
yes

nicHoch committed:
better logging

add version and product filter

fix call error

versioning

move esa duplicates to folder

add documentation and sub1/2 strategy

fix call error

versioning

move esa duplicates to folder

add documentation and sub1/2 strategy

add BSD request reason LUT handling
TODO: connect to API

connect to platform request reason LUT endpoint

remove data file will auto create

clean up data folders

fix remote data test

remove update from test

debug test

debug test

final test tix

update sub

samaloney commented on stixcore/processing/publish.py:
```diff
@@ -0,0 +1,671 @@
+
+    Parameters
+    ----------
+    args : _type_
+        _description_
+
+    Returns
+    -------
+    _type_
+        _description_
```
looks like auto generated txt?

nicHoch replied:
fixed

samaloney commented on stixcore/processing/tests/test_publish.py:
```diff
@@ -0,0 +1,180 @@
+    assert len(res[PublishResult.PUBLISHED]) == 1
+    # 2 where published as supplement
+    assert len(res[PublishResult.MODIFIED]) == 2
+    # 3 where ignored as the data is the same
+    assert len(res[PublishResult.IGNORED]) == 2
```
comment doesn't match the code not sure if just not updated really important?

samaloney commented on docs/pipelineconfiguration.rst:
```diff
@@ -133,3 +133,44 @@ You might toggle the parameter only for manually restarting the service after yo
+                        with already published files
+  -d DB_FILE, --db_file DB_FILE
+                        Path to the history publishing database
+  -w WAITING_PERIOD, --waiting_period WAITING_PERIOD
+                        how long to wait after last file modification before publishing
+  -v INCLUDE_VERSIONS, --include_versions INCLUDE_VERSIONS
+                        what versions should be published
+  -l INCLUDE_LEVELS, --include_levels INCLUDE_LEVELS
+                        what levels should be published
+  -p INCLUDE_PRODUCTS, --include_products INCLUDE_PRODUCTS
```
would be nice to show the examples of valid values? Is this a copy past of the running the command with `-h` if so maybe add to the actual argparse config?

nicHoch replied:
will fix

samaloney reviewed (approved):
Wow this does a lot. I think the code could do with being refactored to:
1.  be a bit more pythonic 
2. improve separation of concerns publish.py does everything updating, copying, history etc

but it's working and we need it so I think this could be done at a later stage if there ever time.

nicHoch committed:
enhance after review
